File: Java/프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        List<Integer>[] adj = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) adj[i] = new ArrayList<>();

        for (int[] e : roads) {
            int a = e[0], b = e[1];
            adj[a].add(b);
            adj[b].add(a);
        }

        int[] dist = new int[n + 1];
        Arrays.fill(dist, -1);

        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.add(destination);
        dist[destination] = 0;

        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int nxt : adj[cur]) {
                if (dist[nxt] == -1) {
                    dist[nxt] = dist[cur] + 1;
                    q.add(nxt);
                }
            }
        }

        int[] ans = new int[sources.length];
        for (int i = 0; i < sources.length; i++) {
            ans[i] = dist[sources[i]];
        }
        return ans;
    }
}

File: Python3/프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque

def solution(n, roads, sources, destination):
    adj = [[] for _ in range(n + 1)]
    for a, b in roads:
        adj[a].append(b)
        adj[b].append(a)

    dist = [-1] * (n + 1)
    q = deque([destination])
    dist[destination] = 0

    while q:
        cur = q.popleft()
        for nxt in adj[cur]:
            if dist[nxt] == -1:
                dist[nxt] = dist[cur] + 1
                q.append(nxt)

    return [dist[s] for s in sources]
